### PR TITLE
Docker and things

### PR DESCRIPTION
**Pull request by jwaldrip**

### CODE REVIEW

File: src/templates/app/LICENSE [hunk 1]
```diff
@@ -1,21 +0,0 @@
-The MIT License (MIT)
```
Comment (drujensen) [inline src/templates/app/LICENSE]: :+1: `crystal init app {name}` generates this license as well, but I agree auto generating one may not be the best idea.

File: src/templates/app/shard.yml.ecr [hunk 2]
```diff
@@ -4,13 +4,27 @@ version: 0.1.0
 authors:
   - Amber <contact@ambercr.io>
 
-crystal: 0.23.0
+crystal: <%= Crystal::VERSION %>
```
Comment (drujensen) [inline src/templates/app/shard.yml.ecr]: excellent! 🎆

File: src/templates/app/config/deploy/Dockerfile.ecr [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+ARG DEBIAN_FRONTEND=noninteractive
+RUN apt-get update -qq && apt-get install -y --no-install-recommends libpq-dev libsqlite3-dev libmysqlclient-dev libreadline-dev
+
+# Build the app
+ENV PATH /app/bin:$PATH
+WORKDIR /app
+COPY shard.yml shard.lock /app/
+COPY . /app
+RUN shards build
+
```
Comment (drujensen) [inline src/templates/app/config/deploy/Dockerfile.ecr]: What do you think about removing everything but the binary and `public`?

File: src/templates/app/Dockerfile.ecr [hunk 4]
```diff
@@ -1,15 +1,14 @@
-FROM drujensen/crystal:0.23.0
+FROM crystallang/crystal:<%= Crystal::VERSION %>
 
-ENV AMBER_VERSION <%= VERSION %>
+# Install Dependencies
+ARG DEBIAN_FRONTEND=noninteractive
+RUN apt-get update -qq && apt-get install -y --no-install-recommends libpq-dev libsqlite3-dev libmysqlclient-dev libreadline-dev netcat
 
-RUN curl -L https://github.com/amber-crystal/amber-cmd/archive/v$AMBER_VERSION.tar.gz | tar xvz -C /usr/local/share/. && cd /usr/local/share/amber-cmd-$AMBER_VERSION && crystal deps && make
```
Comment (drujensen) [inline src/templates/app/Dockerfile.ecr]: nice!

File: src/templates/app/Dockerfile.ecr [hunk 5]
```diff
@@ -1,15 +1,14 @@
...
+# Build the app
+ENV PATH /app/bin:$PATH
+WORKDIR /app
+COPY shard.yml shard.lock /app/
+RUN shards build amber
```
Comment (drujensen) [inline src/templates/app/Dockerfile.ecr]: brilliant!  What do you think of moving the binary to `/usr/local/bin`?

File: src/templates/app/shard.yml.ecr [hunk 6]
```diff
@@ -4,13 +4,27 @@ version: 0.1.0
...
 
 license: MIT
 
+targets:
+  <%= @name %>:
+    main: src/<%= @name %>.cr
+
+  amber:
```
Comment (drujensen) [inline src/templates/app/shard.yml.ecr]: Do you think this would be confusing for a user to see multiple targets?

File: src/templates/app/LICENSE (hunk 1 above)
Comment (elorest) [inline src/templates/app/LICENSE]: I would argue that generating these licenses are fine. We're an OSS project and we like other OSS projects. If someone wants to delete the license they can. I could probably go either way on this.

File: src/templates/app/config/deploy/Dockerfile.ecr (hunk 3 above)
Comment (elorest) [inline src/templates/app/config/deploy/Dockerfile.ecr]: The current heroku like deploy process is using this file. Modifying it is fine as long as `amber deploy` still works or is modified to work with the new Dockerfile.

File: src/templates/app/shard.yml.ecr (hunk 2 above)
Comment (elorest) [inline src/templates/app/shard.yml.ecr]: 🦎

File: src/templates/app/shard.yml.ecr (hunk 6 above)
Comment (elorest) [inline src/templates/app/shard.yml.ecr]: Why would the target be amber. Amber seems to be a dependency not a target.

File: src/templates/app/shard.yml.ecr [hunk 7]
```diff
@@ -4,13 +4,27 @@ version: 0.1.0
...
+    main: lib/amber_cmd/src/amber_cmd.cr
+
+development_dependencies:
```
Comment (elorest) [inline src/templates/app/shard.yml.ecr]: Why are development deps moved above deps?

File: src/templates/app/shard.yml.ecr [hunk 8]
```diff
@@ -4,13 +4,27 @@ version: 0.1.0
...
+  amber_spec:
+    github: amber-crystal/amber-spec
+
 dependencies:
+  amber_cmd:
```
Comment (elorest) [inline src/templates/app/shard.yml.ecr]: Why is amber cmd a dep for an amber project?

File: src/templates/app/LICENSE (hunk 1 above)
Comment (jwaldrip) [inline src/templates/app/LICENSE]: If the license ends up in the git history it can cause legal issues for the maintainer... best leave it to the end user to set their own terms.

File: src/templates/app/config/deploy/Dockerfile.ecr (hunk 3 above)
Comment (jwaldrip) [inline src/templates/app/config/deploy/Dockerfile.ecr]: We could do this with a multistage build. If all are agreed I can make this change.

File: src/templates/app/Dockerfile.ecr (hunk 5 above)
Comment (jwaldrip) [inline src/templates/app/Dockerfile.ecr]: No need in this case... the $PATH variable does what you need.

File: src/templates/app/shard.yml.ecr (hunk 6 above)
Comment (jwaldrip) [inline src/templates/app/shard.yml.ecr]: moving the amber target into the shard allows me to add the correct version of amber to `./bin` this means it would always be the amber in the dependencies.

File: src/templates/app/shard.yml.ecr (hunk 8 above)
Comment (jwaldrip) [inline src/templates/app/shard.yml.ecr]: Do you want to run amber commands? This is the best pattern to do so. It locks the version of amber command to the one the project supports.

File: src/templates/app/config/deploy/Dockerfile.ecr (hunk 3 above)
Comment (elorest) [inline src/templates/app/config/deploy/Dockerfile.ecr]: @jwaldrip I'm fine with that as long as it can automatically do deployments and redeployments to $5/m DO servers. Deployment should probably be done in a different PR though. Personally I'd say remove this dockerfile change from this PR and address it in deployments.

File: src/templates/app/Dockerfile.ecr (hunk 5 above)
Comment (drujensen) [inline src/templates/app/Dockerfile.ecr]: Aah, missed that.  :+1:

File: src/templates/app/config/deploy/Dockerfile.ecr (hunk 3 above)
Comment (jwaldrip) [inline src/templates/app/config/deploy/Dockerfile.ecr]: Can do.

File: src/templates/app/shard.yml.ecr (hunk 7 above)
Comment (jwaldrip) [inline src/templates/app/shard.yml.ecr]: Just moved it up there because I was working with it. Does it matter?

File: src/templates/app/shard.yml.ecr (hunk 6 above)
Comment (drujensen) [inline src/templates/app/shard.yml.ecr]: Yeah, at first, this was confusing for me to see multiple targets listed and `amber_cmd` as a dependency. 

But having the version of the `amber_cmd` used when the project was built and locking it in, is a major plus.  I think developers will be happy to find `bin/amber` in the project directory and `lib/amber_cmd` as a dependency.  I know I will.

File: src/templates/app/shard.yml.ecr (hunk 8 above)
Comment (elorest) [inline src/templates/app/shard.yml.ecr]: Similar to the rails pattern of `./bin/rails bundle spring` etc... Where are you compiling this?

**Comment by jwaldrip** [comment]
@elorest how is it looking now?

File: spec/init_command_spec.cr [hunk 9]
```diff
@@ -24,34 +24,35 @@ begin
...
-        File.read_lines("#{TESTING_APP}/shard.yml")[20]?.should eq "  pg:"
+        YAML.parse(File.read("#{TESTING_APP}/config/database.yml"))["pg"].should_not be_nil
+        YAML.parse(File.read("#{TESTING_APP}/shard.yml"))["dependencies"]["pg"].should_not be_nil
       end
 
       it "should create app with mysql settings" do
         Amber::CMD::MainCommand.run ["new", TESTING_APP, "-d", "mysql"]
-        File.read_lines("#{TESTING_APP}/config/database.yml").first.should eq "mysql:"
-        File.read_lines("#{TESTING_APP}/shard.yml")[20]?.should eq "  mysql:"
+        YAML.parse(File.read("#{TESTING_APP}/config/database.yml"))["mysql"].should_not be_nil
```
Comment (elorest) [inline spec/init_command_spec.cr]: :)
